### DISCUSSION TITLE
ContextVar with synchronous dependency

### DISCUSSION THREAD

**Discussion by artemholovnia** [fastapi/fastapi#10124] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from contextvars import ContextVar

from fastapi import Depends, FastAPI

test_ctx = ContextVar("test_ctx", default="hello")

app = FastAPI()


def dep():
    def orig():
        test_ctx.set("world")
        return test_ctx.get()

    return Depends(orig)


@app.get("/test", dependencies=(dep(),))
async def test():
    return test_ctx.get()

```

### Description

When i define in dependencies `dep()` function, which returns `Depends` object i expect that returns value of function `test` must be "hello", but it returns "world", as like default value of context var `test_ctx`. **Is it normal behavior?**

### Operating System

macOS

### FastAPI Version

0.99.1

### Pydantic Version

1.10.10

### Python Version

Python 3.10.12

**Comment by Kludex** [comment]
Please provide an example with imports.

**Reply by artemholovnia** [reply]
updated

**Comment by romabozhanovgithub** [comment]
Hey, I have tried to execute your code and here is what I got:
## FastAPI
```python
from contextvars import ContextVar

from fastapi import Depends, FastAPI

test_ctx = ContextVar("test_ctx", default="hello")

app = FastAPI()


def dep():
    def orig():
        test_ctx.set("world")
        return test_ctx.get()

    return Depends(orig)


@app.get("/test", dependencies=(dep(),))
async def test():
    return test_ctx.get()
```
## Response
```json
"hello"
```

Do I miss something?

**Reply by artemholovnia** [reply]
Ok, but response must be "world" against "hello", cause context variable `test_ctx` has been set via DI function `dep` to "world"

**Reply by romabozhanovgithub** [reply]
Oh, ok, maybe you misstyped there?
_which returns Depends object i expect that returns value of function test must be "hello", but it returns "world"_

**Reply by romabozhanovgithub** [reply]
In the dep() function, a function orig() is defined which sets the value of test_ctx to "world" and returns its value. However, this function is not actually called until the test() function is executed.

When the test() function is executed, it retrieves the value of test_ctx using test_ctx.get(). Since test_ctx has not been modified in the current context, it returns the default value of "hello".

## How it works:
```python
from contextvars import ContextVar

# Create a ContextVar object with a default value of 0
counter = ContextVar("counter", default=0)

# Define a function that increments the counter
def increment_counter():
    current_value = counter.get()
    counter.set(current_value + 1)

# Call the function in a new context
with counter.replace(10):
    increment_counter()
    print(counter.get())  # Output: 11

# The counter value is still 0 in the original context
print(counter.get())  # Output: 0
```

**Reply by artemholovnia** [reply]
imho, it is unpredictable behavior, but i saw below what is the problem, the problem was in place when basic function, not async is called, when function calls in extra thread that creates another context for variable and change it whithis this context.
If i change the real dependency function as below
```python
def dep():
    async def orig():
        test_ctx.set("world")
        return test_ctx.get()

    return Depends(orig)
```
it works, cause main dependency function is async function

**Reply by Kludex** [reply]
https://github.com/tiangolo/fastapi/discussions/10124#discussioncomment-6788263

**Comment by Kludex** [comment, marked as answer]
This is a limitation, not a bug.

- You can see the surrounding discussion here: https://github.com/encode/starlette/pull/1258.

I believe it's documented somewhere... If it's not (please check), PR welcome. An admonition note/warning would be cool.
